bluemix/terminal/table.go

```diff
@@ -69,58 +75,152 @@ func (t *PrintableTable) Add(row ...string) {
+	if len(t.rows) == 0 {
+		return []table.ColumnConfig{}
+	}
+
+	colCount := len(t.rows[0])
+	var (
+		widestColIndicies []int
+		terminalWidth     = terminalWidth()
+		// total amount padding space that a row will take up
+		totalPaddingSpace = (colCount - 1) * minSpace
+		remainingSpace    = terminalWidth - totalPaddingSpace
+		// the estimated max column width by dividing the remaining space evenly across the columns
+		maxColWidth = (terminalWidth - totalPaddingSpace) / colCount
+	)
+	columnConfig := make([]table.ColumnConfig, len(t.maxSizes))
+
+	for i := range columnConfig {
+		columnConfig[i] = table.ColumnConfig{
+			AlignHeader: text.AlignLeft,
+			Align:       text.AlignLeft,
+			WidthMax:    maxColWidth,
+			Number:      i + 1,
+		}
+
+		// assuming the table has headers: store columns with wide content where the max width may need to be adjusted
+		// using the remaining space
+		if t.maxSizes[i] > maxColWidth && (len(t.headers) > 0 && isWideColumn(t.headers[i])) {
+			widestColIndicies = append(widestColIndicies, i)
+		} else if t.maxSizes[i] < maxColWidth {
+			// use the max column width instead of the estimated max column width
+			// if it is shorter
+			columnConfig[i].WidthMax = t.maxSizes[i]
+			remainingSpace -= t.maxSizes[i]
+		} else {
+			remainingSpace -= maxColWidth
+		}
 	}
 
-	for _, line := range t.rows {
-		t.printRow(line)
+	// if only one wide column use the remaining space as the max column width
+	if len(widestColIndicies) == 1 {
+		idx := widestColIndicies[0]
+		columnConfig[idx].WidthMax = remainingSpace
 	}
 
-	t.rows = [][]string{}
-}
+	// if more than one wide column, spread the remaining space between the columns
+	if len(widestColIndicies) > 1 {
+		remainingSpace /= len(widestColIndicies)
+		for _, columnCfgIdx := range widestColIndicies {
+			columnConfig[columnCfgIdx].WidthMax = remainingSpace
+		}
 
-func (t *PrintableTable) calculateMaxSize(row []string) {
-	for index, value := range row {
-		cellLength := runewidth.StringWidth(Decolorize(value))
-		if t.maxSizes[index] < cellLength {
-			t.maxSizes[index] = cellLength
+		origRemainingSpace := remainingSpace
+		moreRemainingSpace := origRemainingSpace % len(widestColIndicies)
+		if moreRemainingSpace != 0 {
+			columnConfig[0].WidthMax += moreRemainingSpace
 		}
 	}
+
+	return columnConfig
 }
 
-func (t *PrintableTable) printHeader() {
-	output := ""
-	for col, value := range t.headers {
-		output = output + t.cellValue(col, HeaderColor(value))
+func (t *PrintableTable) createPrettyRowsAndHeaders() (headerRow table.Row, rows []table.Row) {
+	for _, header := range t.headers {
+		headerRow = append(headerRow, header)
 	}
-	fmt.Fprintln(t.writer, output)
-}
 
-func (t *PrintableTable) printRow(row []string) {
-	output := ""
-	for columnIndex, value := range row {
-		if columnIndex == 0 {
-			value = TableContentHeaderColor(value)
+	for i := range t.rows {
+		var row, emptyRow table.Row
+		for i, cell := range t.rows[i] {
```

Comment:
Variable name `i` is used twice in this context. Let's also rename these so that they indicate what they are doing. For example, the 1st i is a rowIndex and the 2nd a cellIndex it appears.

Reply:
Yeah good catch. Let me update